Update heroku deploy docs (#25457)
backstage/backstage: docs/deployment/heroku.md

description: How to deploy Backstage to Heroku
---

Heroku is a Platform as a Service (PaaS) designed to simplify application deployment.

## Create App

> Start with an exisiting Backstage app or follow the [getting started guide](https://backstage.io/docs/getting-started/) to create a new one:

Reviewer: ```suggestion
Starting with an existing Backstage app or follow the [getting started guide](https://backstage.io/docs/getting-started/) to create a new one:
```

```shell
$ npx @backstage/create-app@latest
  Enter a name <your-app>
```

Install the
[Heroku CLI](https://devcenter.heroku.com/articles/heroku-cli) and create a new Heroku app:

```shell
$ cd your-app/
$ heroku apps:create <your-app>
```

## Domain

Get Heroku app URL:
```shell
$ heroku domains -a <your-app>
<your-app-123>.herokuapp.com
```

Update `baseUrl` and `port` in `app-config.yaml`. The core [app-backend plugin](https://www.npmjs.com/package/@backstage/plugin-app-backend) allows a single Heroku app to serve the frontend and backend:
```yaml
app:
  baseUrl: https://<your-app-123>.herokuapp.com

backend:
  baseUrl: https://<your-app-123>.herokuapp.com
  listen:
    port:
      $env: PORT
      # The $PORT environment variable is a feature of Heroku
      # https://devcenter.heroku.com/articles/dynos#web-dynos
```

## Build Script

Add a build script in `package.json` to compile frontend during deployment:
```json
"scripts": {

>   "build": "yarn build:backend --config ../../app-config.yaml"

Reviewer: ```suggestion
  "build": "yarn build:backend --config ../../app-config.yaml --config app-config.production.yaml"
```

```

> ## Start Command

Reviewer: Is there a reason we aren't suggesting running in a Docker container? We do this right now with the Demo site. I think perhaps we should keep both methods.

Author: I have added the docker instructions.

description: How to deploy Backstage to Heroku
---

Heroku is a Platform as a Service (PaaS) designed to simplify application deployment.

> [![Deploy](https://www.herokucdn.com/deploy/button.svg)](https://heroku.com/deploy?template=https://github.com/chap/backstage-button)

Reviewer: If we can remove this that would be ideal, just leaves us open to something going unmaintained that can be easily missed.

Author: No problem, removed

```

## Database

Provision a [Heroku Postgres](https://elements.heroku.com/addons/heroku-postgresql) database:

```shell
$ heroku addons:create heroku-postgresql -a <your-app>
```

Update `database` in `app-config.production.yaml`:

```yaml
backend:
  database:
    client: pg
    pluginDivisionMode: schema
    ensureExists: false
    ensureSchemaExists: true
    connection: ${DATABASE_URL}
```

Allow postgres self-signed certificates:

```shell
$ heroku config:set PGSSLMODE=no-verify -a <your-app>
```

## Deployment

Commit changes and push to Heroku to build and deploy:

```shell
$ git add Procfile && git commit -am "configure heroku"
$ git push heroku main
```

View the app in the browser:

```shell
$ heroku open -a <your-app>
```

View logs:

```shell
$ heroku heroku -a <your-app>
```

## Docker

As an alternative to git deploys, Heroku also [supports container images](https://devcenter.heroku.com/articles/container-registry-and-runtime).

Login to Heroku's container registry:

```shell
$ heroku container:login
```

Configure the Heroku app to run a container image:

```shell
$ heroku stack:set container -a <your-app>
```

> Locally run the [host build command](https://backstage.io/docs/deployment/docker/#host-build):

Reviewer: ```suggestion
Locally run the [host build commands](https://backstage.io/docs/deployment/docker/#host-build), they must be run whenever you are going to publish a new image:
```

Author: fixed

```shell

> $ yarn build:backend --config ../../app-config.yaml --config ../../app-config.production.yaml

Reviewer: ```suggestion
yarn install --frozen-lockfile
yarn tsc
yarn build:backend --config ../../app-config.yaml --config ../../app-config.production.yaml
```

Author: fixed

```

> Build, push, and release the container image:

Reviewer: ```suggestion
Build, push, and release the container image to the `web` worker:
```
